util/logtools/encounter_tools.ts

```diff
@@ -0,0 +1,283 @@
+import ContentType from '../../resources/content_type';
+import NetRegexes from '../../resources/netregexes';
+import { UnreachableCode } from '../../resources/not_reached';
+import ZoneInfo from '../../resources/zone_info';
+import { NetMatches, NetAnyMatches } from '../../types/net_matches';
+import { CactbotBaseRegExp } from '../../types/net_trigger';
+import { commonReplacement, syncKeys } from '../../ui/raidboss/common_replacement';
+
+// TODO: add some error checking that a zone has been found before a fight.
+// This can happen on partial logs.
+
+type ZoneEncInfo = {
+  name?: string;
+  zoneId?: number;
+  startLine?: string;
+  endLine?: string;
+  startTime?: Date;
+  endTime?: Date;
+};
+
+type FightEncInfo = {
+  name?: string;
+  zoneId?: number;
+  startLine?: string;
+  endLine?: string;
+  startTime?: Date;
+  endTime?: Date;
+  endType?: string;
+  sealName?: string;
+};
+
+export class EncounterFinder {
+  currentZone: ZoneEncInfo = {};
+  currentFight: FightEncInfo = {};
+  currentSeal?: string;
+  zoneInfo?: typeof ZoneInfo[number];
+
+  haveWon = false;
+  haveSeenSeals = false;
+
+  regex: {
+    changeZone: CactbotBaseRegExp<'ChangeZone'>;
+    cactbotWipe: CactbotBaseRegExp<'GameLog'>;
+    win: CactbotBaseRegExp<'ActorControl'>;
+    wipe: CactbotBaseRegExp<'ActorControl'>;
+    commence: CactbotBaseRegExp<'ActorControl'>;
+    playerAttackingMob: CactbotBaseRegExp<'Ability'>;
+    mobAttackingPlayer: CactbotBaseRegExp<'Ability'>;
+  };
+
+  sealRegexes: Array<CactbotBaseRegExp<'GameLog'>> = [];
+  unsealRegexes: Array<CactbotBaseRegExp<'GameLog'>> = [];
+
+  initializeZone(): void {
+    this.currentZone = {};
```

Comment:
It occurs to me, maybe we should re-initialize `this.zoneInfo` here as well? In theory, `skipZone` should be called only in places where `zoneInfo` is correctly populated, but would it be bad to be explicit about it?

Reply:
Yeah, I think that makes sense.  Good idea!